File: src/main/scala/com/lp/scala/demo/datastream/source/KafkaSourceApp.scala
package com.lp.scala.demo.datastream.source

import com.lp.scala.demo.datastream.trigger.CustomProcessTimeTrigger
import com.lp.scala.demo.utils.ConfigUtils
import org.apache.flink.api.common.functions.RichMapFunction
import org.apache.flink.api.common.restartstrategy.RestartStrategies
import org.apache.flink.api.common.serialization.SimpleStringSchema
import org.apache.flink.runtime.jobgraph.JobVertex
import org.apache.flink.streaming.api.environment.CheckpointConfig
import org.apache.flink.streaming.api.scala.StreamExecutionEnvironment
import org.apache.flink.streaming.api.windowing.time.Time
import org.apache.flink.streaming.api.{CheckpointingMode, TimeCharacteristic}
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaConsumer


/**
  * <p/> 
  * <li>Description: kafka消费者测试,自定义触发器</li>
  * <li>@author: panli0226@sina.com</li> 
  * <li>Date: 2019-05-07 22:31</li> 
  */
object KafkaSourceApp {

  def main(args: Array[String]): Unit = {

    //获取flink流式运行环境
    val env = StreamExecutionEnvironment.getExecutionEnvironment

    //设置最少处理一次语义和恰好一次语义
    env.enableCheckpointing(20000, CheckpointingMode.AT_LEAST_ONCE)

    //checkpointing可以分开设置
    //env.enableCheckpointing(20000)
    //env.getCheckpointConfig.setCheckpointingMode(CheckpointingMode.EXACTLY_ONCE)

    //设置checkpoint目录 state存储路径

    //env.setStateBackend(new FsStateBackend("/hdfs/checkpoint"));
    //设置checkpoint的清除策略
    env.getCheckpointConfig.enableExternalizedCheckpoints(CheckpointConfig.ExternalizedCheckpointCleanup.RETAIN_ON_CANCELLATION)

    /**
      * 设置重启策略/5次尝试/每次重试间隔50s
      */
    env.setRestartStrategy(RestartStrategies.fixedDelayRestart(5, 50000))

    //设置flink以身为时间为基准作，处理事件
    env.setStreamTimeCharacteristic(TimeCharacteristic.ProcessingTime)
    val kafkaConfig = ConfigUtils.apply("string")

    val kafkaConsumer = new FlinkKafkaConsumer(kafkaConfig._1, new SimpleStringSchema(), kafkaConfig._2)
      .setStartFromLatest()

    import org.apache.flink.api.scala._

    val stream = env
      .addSource(kafkaConsumer)
      .setParallelism(1)
      .map(new RichMapFunction[String, Int] {
        override def map(value: String): Int = {
          Integer.valueOf(value)
        }
      })
      .timeWindowAll(Time.seconds(20))
      .trigger(new CustomProcessTimeTrigger) //10个元素触发一次计算

    stream.sum(0).print()

    // 获取JobGraph
//    val vertices = env.getStreamGraph.getJobGraph.getVertices
//    import scala.collection.JavaConversions._
//    for (vertex <- vertices) {
//      System.out.println("=====>" + vertex.getName)
//      System.out.println("=====>" + vertex.getID)
//    }
    env.execute("KafkaSourceTest")


  }

}

File: src/main/java/com/lp/java/demo/datastream/source/KafkaSourceApp.java
package com.lp.java.demo.datastream.source;

import com.lp.java.demo.datastream.BaseStreamingEnv;
import com.lp.java.demo.base.IBaseRunApp;
import com.lp.java.demo.commons.po.config.JobConfigPo;
import com.lp.java.demo.commons.po.config.KafkaConfigPo;
import com.lp.java.demo.datastream.windows.trigger.CustomProcessingTimeTrigger;
import org.apache.flink.api.common.functions.RichMapFunction;
import org.apache.flink.api.common.serialization.SimpleStringSchema;
import org.apache.flink.streaming.api.datastream.AllWindowedStream;
import org.apache.flink.streaming.api.windowing.assigners.TumblingProcessingTimeWindows;
import org.apache.flink.streaming.api.windowing.time.Time;
import org.apache.flink.streaming.api.windowing.windows.TimeWindow;
import org.apache.flink.streaming.connectors.kafka.FlinkKafkaConsumer;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;


/**
 * <p/>
 * <li>Description: kafka消费者测试,自定义触发器</li>
 * <li>@author: panli0226@sina.com</li>
 * <li>Date: 2019-05-07 22:31</li>
 */
public class KafkaSourceApp extends BaseStreamingEnv<String> implements IBaseRunApp {

    private final static Logger log = LoggerFactory.getLogger(KafkaSourceApp.class);

    @Override
    public Integer setDefaultParallelism() {
        return 1;
    }

    @Override
    public void doMain() throws Exception {
        // 指定kafka topic和序列化方式
        FlinkKafkaConsumer<String> kafkaConsumer =
                getKafkaConsumer(KafkaConfigPo.stringTopic, new SimpleStringSchema());

        AllWindowedStream<Integer, TimeWindow> stream =
                env
                        .addSource(kafkaConsumer)
                        .map(new String2Integer())
                        .windowAll(TumblingProcessingTimeWindows.of(Time.seconds(10)))
//                        .trigger(ProcessingTimeTrigger.create());
                        // 自定义触发器在满足条件时会出发,窗口结束时也会触发
                        .trigger(CustomProcessingTimeTrigger.create());


        // 聚合所有窗口
        stream.sum(0).print();


        // 获取JobGraph 调用 getStreamGraph 会清除 transformations 避免和 execute并行使用
//        Iterable<JobVertex> vertices = stream.getExecutionEnvironment().getStreamGraph().getJobGraph().getVertices();
//        for (JobVertex vertex : vertices) {
//            log.info("vertex ====>《Name:》 {} 《Id:》 {}", vertex.getName(), vertex.getID());
//        }

        env.execute(JobConfigPo.jobNamePrefix + KafkaSourceApp.class.getName());


    }


    private static class String2Integer extends RichMapFunction<String, Integer> {
        private static final long serialVersionUID = 1180234853172462378L;

        @Override
        public Integer map(String event) throws Exception {

            return Integer.valueOf(event);
        }
    }

}
